pkg/receive/handler.go

```diff
@@ -456,6 +487,21 @@ func (h *Handler) receiveHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// Impose limits only if Receive is in Router or RouterIngestor mode.
+	if h.receiverMode == RouterOnly || h.receiverMode == RouterIngestor {
+		under, err := h.isUnderLimit(ctx, tenant, tLogger)
+		if err != nil {
+			level.Info(tLogger).Log("msg", "error while limiting", "err", err.Error())
+
+		}
+
+		// Fail request fully if tenant has exceeded set limit.
+		if !under {
+			http.Error(w, "tenant is above active series limit", http.StatusTooManyRequests)
```

Comment:
Yes, that is essentially what a client like Prometheus does, it retries up til 2 hours: https://prometheus.io/docs/practices/remote_write/#remote-write-characteristics. So the client would need to retry until their head series metric goes below set limit. And Thanos fails the request fully, so there is no confusion of partial writes.

To answer your comment question, yes this is the approach chosen and the proposal has been updated to reflect that. The other approaches are listed as alternatives. 🙂

Reply:
Yes, I was also confused about what status code would be suitable here. I saw that an explicit config option exists, as you mentioned, which is why I used 429. But maybe there is some better default here? 🤔 

So once a tenant hits their limit, all their remote write requests would fail. So we would not ingest any data, but would still have to partially process the retry requests (essentially till `isUnderLimit`). Once the metric decreases in value to under the limit, we can start ingesting again. I kind of wanted to replicate such an error state in a dedicated e2e test, but haven't been able to figure out how to do so via containers. Will still try to see if there's a clever but accurate way of doing this!  🙂

Reply:
The proposal states that:

> Thus, any remote write request can be failed completely, with a 429 status code and appropriate error message, **if it increases the number of active series above the configured limit for a tenant**

So not all their remote write requests would fail. Those requests that won't create any net-new head series should still be accepted.

Reply:
I don't think there is a good HTTP code for this. 

HTTP code 413 for Entity Too Large doesn't make sense as the "entity" in the request can probably be accepted at a later time. A request that gets a 413 should never be retried.

We have 422 for Unprocessable Entity, but I never seen it being used in cases of limits. 

I think we can stick to the 429, which is the closest to the meaning of the situation here. We can include a `Retry-After` header to help clients stop retrying non-stop and give the servers some rest. 😴

Reply:
> So not all their remote write requests would fail. Those requests that won't create any net-new head series should still be accepted.

That isn't possible in this approach as we don't check how much a remote write request increases the number of head series. But you're right, there's an error in the proposal. Let me update! 🙂

Reply:
> I'm also not sure if returning retry after makes sense because we don't know when in the future the limit will be lifted, given this is not a time-based rate limiting (and besides Prometheus has internal retry mechanism with backoff).

The point of the retry after header in this case is to avoid badly configured Prometheus remote write clients spamming retries. It would work as a self-defense mechanism for Receive, as long as the client respects it (optional maybe?). 

> 429 - this means we won't be able to predict whether client retries on failed requests (Do we want the client to retry in all cases? Are we fine with having this optional? Can we ensure clients are aware of this behavior in receiver?)

Prometheus remote write client already retries on 429 if they are configured to do so, this stays the same. Clients are aware that they might be limited and decide if they want to retry as their data might be accepted later on -- it's a temporary limit. 

I don't see any reason for a different behavior.

Reply:
> The point of the retry after header in this case is to avoid badly configured Prometheus remote write clients spamming retries. It would work as a self-defense mechanism for Receive, as long as the client respects it (optional maybe?).

That feels a bit to me like a misuse of the header. As a client, I'd have legitimate expectation that my request (if it's not a bad request) will go through after specified time limit. But we cannot guarantee that in this case, because we don't know when this will happen. If we want to prevent abuse by 'bad' clients in general, perhaps there are other mechanisms worth exploring, outside of the scope of this PR.

> Prometheus remote write client already retries on 429 if they are configured to do so, this stays the same. Clients are aware that they might be limited and decide if they want to retry as their data might be accepted later on -- it's a temporary limit.

I think I'm just wondering how temporary is the limit. This basically refers back to what we discussed with @saswatamcode above in https://github.com/thanos-io/thanos/pull/5520#discussion_r925146840./ How long until an active series becomes 'free' for a tenant? (Maybe it would be good to document this so users understand better what constitutes an active series and how long it is active). Could it be that in an extreme case, client _A_ might never get their request in, because once the limit drops, it gets filled by another client _X_ and thus  client _A_ will always see `429`? It feels like the clients would be constantly 'fighting' for the few spot. These are still some of the aspects I'm trying to grasp my head around and wonder whether they could pose issues, maybe I just a need a clarification on my understanding of the limits.

Reply:
>  How long until an active series becomes 'free' for a tenant? 

100% we have to document this. I found in [Grafana's docs](https://grafana.com/docs/grafana-cloud/fundamentals/active-series-and-dpm/#active-series) that `For Prometheus, a time series is considered active if new data points have been generated within the last 15 to 30 minutes.`. I couldn't find any official information about in Prometheus' it docs though. 

So the temporary aspect of the limit could range from almost 0 to almost 30 minutes. 

> Could it be that in an extreme case, client A might never get their request in, because once the limit drops, it gets filled by another client X and thus client A will always see 429? It feels like the clients would be constantly 'fighting' for the few spot.

This could happen. It's a common problem of multiple clients under the same rate limiting group being very active.

In a non-distributed, one way to manage this is to have a central function/class/whatever that will receive requests from different sources and weigh in how many from each will be sent to the server. This way it could try to be fair and give each source 50% of the available limit.

In a distributed system it's more difficult to coordinate this. Even more difficult in the remote write context, as clients often are completely isolate from each other.

Reply:
> Could it be that in an extreme case, client A might never get their request in, because once the limit drops, it gets filled by another client X and thus client A will always see 429? It feels like the clients would be constantly 'fighting' for the few spot.

So yes, this can happen with client X and client A both sending remote write requests with tenant headers for the same tenant and one of them essentially getting "starved" due to limiting and having to fight for some HEAD series spots. But that is kind of the point of limiting I think, where we protect Thanos ingestion system health, by blocking requests. 🙂 

If a non-admin persona Thanos user notices that a client of theirs cannot seem to send metrics due to limits being hit continually, they would have a few solutions,

- Request admin persona Thanos user to increase limits so their starved client actually gets a chance to send metrics
- Re-evaluate what metrics are being sent by their clients and maybe apply metric keep/drop configs accordingly
- Load balance remote write requests to give "starved" clients a fair chance

> This could happen. It's a common problem of multiple clients under the same rate limiting group being very active.
In a non-distributed system, one way to manage this is to have a central function/class/whatever that will receive requests from different sources and weigh in how many from each will be sent to the server. This way it could try to be fair and give each source 50% of the available limit.

Thanos Receive is client-agnostic, and only tracks tenants as separate TSDBs need to be created for each of them. How a particular "quota" of HEAD series is being used by a tenant should perhaps be a client-side issue, in this kind of system? 🙂

Reply:
> How a particular "quota" of HEAD series is being used by a tenant should perhaps be a client-side issue, in this kind of system?

It is totally the client's responsibility. What I mean is that if the client is a distributed system, it's non-trivial to ensure fairness in this scenario. There's nothing we can do.

Reply:
I think we have agreement on that it's client's problem, but I'm still focusing on the server side. Imagine we have 10 Prometheus instances writing to us periodically, but we know 9 out those will not be able to get their requests in (in the imagined "race for the free spot" scenario, as I described above) because of the limit, but we'll still be like "yes please, keep spamming us with requests although we will refuse them with very high probability".

And that's where I'm wondering, are we fine with having clients continuously sending us pointless requests? Does this pose a performance issue? Is there a better way to communicate to clients how long to wait / when a series limit is lifted? For these I don't have good answers :upside_down_face:.

Reply:
> And that's where I'm wondering, are we fine with having clients continuously sending us pointless requests? Does this pose a performance issue? Is there a better way to communicate to clients how long to wait / when a series limit is lifted? For these I don't have good answers

Let me look into this a bit more! While we fail requests early when limit is hit, there might still be some performance problems. Maybe we cannot mitigate via server-side alone, so would need to suggest using some client config options like backoff/retries on client-side.

Reply:
Even strictly outside of Thanos, I'm wondering about impact just in the terms of all the overhead of making wasteful HTTP requests from many clients for extended period of time (maybe I'm being overly worried :shrug:).

With client side, that could work, but if we consider that we are moving into territory where we do not control the clients, it's harder to enforce such configs (otherwise we could just force relabel configuration on the client and we would not need series limits after all :slightly_smiling_face:).

Anyways, I don't want this to lead to a never ending cycle of conversation. Let's see what will be @saswatamcode's findings :bow:.

Reply:
So I think this is how this limit and its recovery will behave under different scenarios (without manual client config changes),

- New tenant on Receive sending requests from clients and remaining under the total limit: Will not have any issues with sending remote write requests
- New tenant on Receive that started going above the limit right from the first remote write request: The first request gets ingested and `prometheus_tsdb_head_series` gauge is incremented, which causes any subsequent remote write requests to be limited. The Receive `prometheus_tsdb_head_series` metric value (can) decrease when Receive tenant TSDB compacts the head block into a new persistent block, which happens every 2 hours. The decrease can also be measured via `prometheus_tsdb_head_series_removed_total`. However, if the clients are only pushing samples of the same series, then the metric may not decrease at all. If the clients have `retry_on_http_429` enabled, they will keep retrying their remote write requests during those 2 hours otherwise they will fail all their remote write requests. How many times they will retry (if enabled) each request is dependent on the `min/max_backoff` options which by default are 30ms to 5s. As this tenant has started going above the limit right from the beginning, it would have to wait for a full 2 hours before getting its limit "quota" back. On the client side you can see how this would limiting would fail or retry remote write requests with metrics like `prometheus_remote_storage_retried_samples_total` or `prometheus_remote_storage_failed_samples_total`.
- Old tenant on Receive that has started going above the limit: If a tenant which was under the set limit for some time, started going above their limit (maybe due to a new client being added/an existing client pushing more data), their remote write requests would start to be limited. The client retry semantics & frequency would be the same as mentioned in the point above. How Receive `prometheus_tsdb_head_series` increases and decreases would also be the same. However, the duration after which this tenant can get its quota back is variable. If the Receive tenant TSDB head block was about to reach block length (`chunkRange * 3/2`), then the tenant can get its quota back in maybe a minute. However, if the head block was just compacted, it might take 2 hours.

I hope this makes it a bit clearer. Feel free to point out if I'm wrong somewhere, or if something you thought could be done differently here. 🙂

Reply:
Thanks for going above and beyond to test this, I appreciate it @saswatamcode :bow:. I think it roughly confirms how I imagined this would work. Do you think it is worth it to document this and to explain behavior this behavior of clients? Then users can also make informed decision whether they want to enable the retry on `429` option on Prometheus side or not.

Reply:
Will add a section to docs!